src/processor/java/org/truffleruby/processor/BuildInformationProcessor.java

```diff
@@ -0,0 +1,119 @@
+        return SourceVersion.latest();
+    }
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnvironment) {
+        if (!annotations.isEmpty()) {
+            for (Element element : roundEnvironment.getElementsAnnotatedWith(PopulateBuildInformation.class)) {
+                try {
+                    processBuildInformation((TypeElement) element);
+                } catch (IOException e) {
+                    processingEnv.getMessager().printMessage(Kind.ERROR, e.getClass() + " " + e.getMessage(), element);
+                }
+            }
+        }
+
+        return true;
+    }
+
+    private void processBuildInformation(TypeElement element) throws IOException {
+        final PackageElement packageElement = (PackageElement) element.getEnclosingElement();
+        final String packageName = packageElement.getQualifiedName().toString();
+
+        final String qualifiedName = element.getQualifiedName().toString();
+
+        if (!processed.add(qualifiedName)) {
+            // Already processed, do nothing. This seems an Eclipse bug.
+            return;
+        }
+
+        final JavaFileObject output = processingEnv.getFiler().createSourceFile(qualifiedName + SUFFIX, element);
+
+        try (PrintStream stream = new PrintStream(output.openOutputStream(), true, "UTF-8")) {
+            stream.println("package " + packageName + ";");
+            stream.println();
+            stream.println("public class " + element.getSimpleName() + SUFFIX + " implements " + element.getSimpleName() + " {");
+            stream.println();
+            stream.println("    public static final " + element.getSimpleName() + " INSTANCE = new " + element.getSimpleName() + SUFFIX + "();");
+            stream.println();
+
+            for (Element e : element.getEnclosedElements()) {
+                if (e instanceof ExecutableElement) {
+                    final String value;
+
+                    switch (e.getSimpleName().toString()) {
+                        case "getRevision":
+                            value = getRevision();
+                            break;
+                        case "getCompileDate":
+                            value = getCompileDate();
+                            break;
+                        default:
+                            throw new UnsupportedOperationException(e.getSimpleName() + " method not understood");
+                    }
+
+                    stream.println("    @Override");
+                    stream.println("    public String " + e.getSimpleName() + "() {");
+                    stream.println("        return \"" + value + "\";");
+                    stream.println("    }");
+                }
+            }
+
+            stream.println();
+            stream.println("}");
+        }
+    }
+
+    public String getRevision() throws IOException {
+        final Process git = Runtime.getRuntime().exec("git rev-parse --short HEAD");
```

Comment:
I don't know if we're always in the correct directory to run this. I'll see what happens in CI, and I'll try it in the GraalVM build as well, and we'll just have to see what happens.

Reply:
You could get the location of the `.jar` containing this class file, but that would still not be the root directory.
Using the `ruby home` would be best, but that might be difficult to determine at `javac` build time.

Reply:
We can CD to the directory containing the annotation class which is actually running, and run `git` from there, as long `mxbuild` is separate for each project.

Reply:
I'm doing that now.

Reply:
It doesn't really matter, does it? Unless you're worried about being in a git submodule, git works fine in subdirectories.